**Reto 2 Arboles de Decisión**

In [11]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt #Graficador
from sklearn.tree import DecisionTreeClassifier

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1.Cargar los datos

In [12]:
nxl='/content/drive/MyDrive/Analitica de Negocios/2. BD1_Travel_Data_int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB=XDB.dropna() #Borra las celdas vacias
XDB.head(100)



College  CreditCard  FoodSpend  Income  TravelSpend
0         1           1    5472.43   49150       827.40
1         0           1    9130.73   47806       863.55
2         0           1    4450.67   46050      1180.05
3         0           1    5020.72   42600       755.70
4         0           1    3408.11   52092       851.40
..      ...         ...        ...     ...          ...
95        0           0    2019.98   65982      4338.75
96        0           0    7046.11   53986      1851.60
97        0           0    5180.13   53986      1646.70
98        0           0    3266.44   63962      1641.60
99        1           1    2082.23   49004      1564.95

[100 rows x 5 columns]

In [13]:
#Seleccion de las variables a trabajar
XD=np.array(XDB[['College','FoodSpend','Income','TravelSpend']]) #Datos de Entrada
print(XD)
yd=np.array(XDB[['CreditCard']]) #Datos de Salida o datos que el modelo va a pronosticar
print(yd)

[[1.00000e+00 5.47243e+03 4.91500e+04 8.27400e+02]
 [0.00000e+00 9.13073e+03 4.78060e+04 8.63550e+02]
 [0.00000e+00 4.45067e+03 4.60500e+04 1.18005e+03]
 ...
 [0.00000e+00 3.58586e+03 4.22500e+04 2.25195e+03]
 [1.00000e+00 1.83395e+03 6.02480e+04 2.00025e+03]
 [0.00000e+00 6.20597e+03 5.85200e+04 3.66750e+03]]
[[1]
 [1]
 [1]
 [1]
 [1]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [1]
 [1]
 [1]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [1]
 [0]
 [1]
 [1]
 [0]
 [1]
 [0]
 [1]
 [1]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [1]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [1]
 [0]
 [1]
 [0]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0

2. Creación del Modelo del arbol de decisión

In [14]:
mar=DecisionTreeClassifier(criterion='gini', max_depth=4) #4 variables de entrada
mar.fit(XD,yd)
ydp=mar.predict(XD)

from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
print(cm)

#Evaluación del modelo
VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]

Ex=(VN+VP)/(VN+VP+FN+FP)  #Que porcentaje de los datos pronostico correctamente
print("La exactitud alcanzada por el modelo es: ",Ex)

Se=VP/(FN+VP) #Como le fue con los positivos
print("La sensibilidad alcanzada por el modelo es: ",Se)

Es=VN/(VN+FP) #Como le fue con los negativos
print("La especificidad alcanzada por el modelo es: ",Es)


[[334   9]
 [ 42 115]]
La exactitud alcanzada por el modelo es:  0.898
La sensibilidad alcanzada por el modelo es:  0.732484076433121
La especificidad alcanzada por el modelo es:  0.9737609329446064


3. Graficación del arbol

In [15]:
#Librerias para gráficar el arbol
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data

vs=['College','FoodSpend','Income','TravelSpend'] #Títulos de los nodos del Arbol
dot_graph=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_graph)
graph.write_png('arbolReto.png')

True

4.Pronostico de los personajes especificos/Nuevos solicitantes

In [16]:
XDB2=pd.read_excel(nxl,sheet_name=1)
XDB2=XDB2.dropna() #Borra las celdas vacias
XDB2.head()

College  FoodSpend  Income  TravelSpend
0        0    2892.90   65982       2151.6
1        1    6017.66   53986       2002.5

In [18]:
XD2= np.array(XDB2[['College','FoodSpend','Income','TravelSpend']]) #Datos de entrada
print(XD2)

#Pronosticos
ydp2=mar.predict(XD2)
print(ydp2)
print("El número de clientes que usaron la tarjeta es:" ,np.sum(ydp2))
print("El numero total de Datos es:" , len(ydp2))
print("El numero de clientes que no usaron la tarjeta es:" , len(ydp2)-np.sum(ydp2))

[[0.00000e+00 2.89290e+03 6.59820e+04 2.15160e+03]
 [1.00000e+00 6.01766e+03 5.39860e+04 2.00250e+03]]
[0 0]
El número de clientes que usaron la tarjeta es: 0
El numero total de Datos es: 2
El numero de clientes que no usaron la tarjeta es: 2


In [19]:
porcentajesi=343/500
porcentajeno=157/500
print("El porcentaje de personas que usaron la tarjeta es de:",porcentajesi)
print("El porcentaje de personas que no usaron la tarjeta es de:",porcentajeno)

El porcentaje de personas que usaron la tarjeta es de: 0.686
El porcentaje de personas que no usaron la tarjeta es de: 0.314


**Análisis de Resultados**

De acuerdo con la base de datos suministrada y el nodo base del arbol de decision se puede observar un total de 500 datos, divididos en 343 Personas que usaron la tarjeta de credito y un total de 157 Personas que no la usaron, esto representa el 68.6% y el 31.4% respectivamente. De acuerdo con el pronostico realizado por el modelo este alcanzo una exactitud de 89.8% aproximadamente lo cual es un valor muy bueno, el modelo tambien alcanzo un 73.2% de sensibilidad aproximadamente lo cual significa que acerto en gran cantidad la clasificacion de los datos, y tambien logro un 97.3 de especificidad.
Del modelo se pueden destacar varios nodos puros los cuales son:

- "[PreNeg:1,PreApr:0]" . Este nodo tiene como reglas: TravelSpend <= 1567.87 AND TravelSpend <= 1188.0 AND FoodSpend <= 4990.73 AND FoodSpend > 4972.55

- "[PreNeg:0,PreApr:47]". Este nodo tiene como reglas: TravelSpend <= 1567.87 AND TravelSpend <= 1188.0 AND FoodSpend > 4990.73

- "[PreNeg:108,PreApr:0]". Este nodo tiene como reglas: TravelSpend > 1567.875 AND Income > 59098.0

Con respecto a los personajes especificos a anlizar se puede decir segun el modelo que ninguno de estos 2 utilizo la tarjeta de credito para realizar los pagos de viajes y de alimentos.